golos.publication/golos.publication.cpp

```diff
@@ -435,15 +442,64 @@ void publication::close_message(structures::mssgid message_id) {
             eosio_assert(state.funds.amount >= 0, "LOGIC ERROR! publication::payrewards: state.funds < 0");
         }
 
-        auto new_rshares = WP(state.rshares) - wdfp_t(FP(mssg_itr->state.netshares));
-        auto new_rsharesfn = WP(state.rsharesfn) - wdfp_t(sharesfn);
+        auto rshares = wdfp_t(FP(mssg_itr->state.netshares));
+        auto new_rshares = WP(state.rshares) - rshares;
+
+        auto rsharesfn = wdfp_t(sharesfn);
+        auto new_rsharesfn = WP(state.rsharesfn) - rsharesfn;
 
         eosio_assert(new_rshares >= 0, "LOGIC ERROR! publication::payrewards: new_rshares < 0");
         eosio_assert(new_rsharesfn >= 0, "LOGIC ERROR! publication::payrewards: new_rsharesfn < 0");
 
         state.rshares = new_rshares.data();
         state.rsharesfn = new_rsharesfn.data();
+        
+        mssg_reward.rshares = rshares.data();
+        mssg_reward.rsharesfn = rsharesfn.data();
+        mssg_reward.funds.amount = payout; 
+    }
+
+    message_table.modify(mssg_itr, name(), [&]( auto &item ) {
+            item.closed = true;
+            item.mssg_reward = mssg_reward;
+        });
+
+    bool pool_erased = false;
+    state.msgs--;
+    if(state.msgs == 0) {
+        if(pool != --pools.end()) {
+            send_poolerase_event(*pool);
+
+            pools.erase(pool);
+            pool_erased = true;
+        }
     }
+    if(!pool_erased)
+        pools.modify(pool, _self, [&](auto &item) {
+            item.state = state;
+            send_poolstate_event(item);
+        });
+
+    transaction trx;
+    trx.actions.emplace_back(action{permission_level(_self, config::active_name), _self, "paymssgrwrd"_n, std::make_tuple(payout, message_id, state)});
+    trx.delay_sec = 0;
+    trx.send((static_cast<uint128_t>(mssg_itr->id) << 64) | message_id.author.value, _self);
+}
+
+void publication::paymssgrwrd(int64_t payout, structures::mssgid message_id, structures::poolstate state) {
```

Comment:
As I can see you doesn't use poolstate in reward paying (you need only `state.funds.symbol`). Don't pass all `poolstate` structure to `paymssgrwrd` action. It's better to pass `payout` as asset instead of `uint64_t` and do it through `mssg_reward` field of message_object (not parameters for action).